src/simulators/ch/ch_runner.hpp

```diff
@@ -0,0 +1,661 @@
+  void apply_z(uint_t qubit, uint_t rank);
+  //Methods for non-clifford gates, including parameters for the Sampling
+  void apply_t(uint_t qubit, double r, int rank);
+  void apply_tdag(uint_t qubit, double r, int rank);
+  void apply_u1(uint_t qubit, complex_t lambda, double r, int rank);
+  void apply_ccx(uint_t control_1, uint_t control_2, uint_t target, uint_t branch, int rank);
+  void apply_ccz(uint_t control_1, uint_t control_2, uint_t target, uint_t branch, int rank);
+  //Measure a Pauli projector on each term in the decomposition and update their coefficients
+  // omega.
+  void apply_pauli_projector(const std::vector<pauli_t> &generators);
+  inline void apply_pauli_projector(const std::vector<pauli_t> &generators, uint_t rank);
+  //Routine for Norm Estimation, thin wrapper for the CHSimulator method that uses AER::RngEngine
+  //to set up the estimation routine.
+  double NormEstimation(uint_t n_samples, AER::RngEngine &rng);
+  double NormEstimation(uint_t n_samples, std::vector<pauli_t> generators, AER::RngEngine &rng);
+
+  //Metropolis Estimation for sampling from the output distribution
+  uint_t MetropolisEstimation(uint_t n_steps, AER::RngEngine &rng);
+  std::vector<uint_t> MetropolisEstimation(uint_t n_steps, uint_t n_shots, AER::RngEngine &rng);
+  //Efficient Sampler for the output distribution of a stabilizer state
+  uint_t StabilizerSampler(AER::RngEngine &rng);
+  std::vector<uint_t> StabilizerSampler(uint_t n_shots, AER::RngEngine &rng);
+  //Utilities for the state-vector snapshot.
+  complex_t amplitude(uint_t x_measure);
+  void state_vector(std::vector<complex_t> &svector, AER::RngEngine &rng);
+
+};
+
+//=========================================================================
+// Implementation
+//=========================================================================
+
+Runner::Runner(uint_t num_qubits)
+{
+  initialize(num_qubits);
+}
+
+void Runner::initialize(uint_t num_qubits)
+{
+  if (states.size() > 0)
+  {
+    states.clear();
+  }
+  if (coefficients.size() > 0)
+  {
+    coefficients.clear();
+  }
+  n_qubits = num_qubits;
+  chi = 1;
+  n_omp_threads = 1;
+  states = std::vector<chstabilizer_t>(1, chstabilizer_t(num_qubits));
+  coefficients.push_back(complex_t(1.,0.));
+}
+
+void Runner::initialize_decomposition(uint_t n_states)
+{
+  chi = n_states;
+  states.reserve(chi);
+  coefficients.reserve(chi);
+  if(states.size() > 1 || coefficients.size() > 1)
+  {
+    throw std::runtime_error(std::string("CHSimulator::Runner was initialized without") + 
+                             std::string("being properly cleared since the last ") +
+                             std::string("experiment."));
+  }
+  chstabilizer_t base_sate(states[0]);
+  complex_t coeff(coefficients[0]);
+  for(uint_t i=1; i<chi; i++)
+  {
+    states.push_back(base_sate);
+    coefficients.push_back(coeff);
+  }
+}
+
+void Runner::initialize_omp(uint_t n_threads, uint_t threshold_rank)
+{
+  if(n_threads == 0)
```

Comment:
There's a ternary operator in C++ that is very handy for this simple situations:
```c++
n_omp_threads = (n_threads == 0 ? 1 : n_threads);
```